[feature] Introduce stacked diagrams

gacarrillor opened the pull request:
This PR extends the QGIS diagram module allowing it to create and display **stacked diagrams**, i.e., diagrams located side by side. We've also taken the opportunity to move the diagram settings from a modal dialog to the **Styling dock widget**.

Stacked diagrams can be configured and **displayed horizontally or vertically**. The **spacing** between the so called subdiagrams can also be configured.

Stacked diagrams allow users to create complex diagrams like **population pyramids**.

Stacked diagrams can show **mixed subdiagram types** (e.g., a pie chart alongside a histogram) and users can add as many subdiagrams as they want. Moreover, subdiagrams can use **different renderer types**: e.g., one subdiagram could use a single category renderer, whereas the other could use a linearly interpolated one.

Subdiagrams can **set their own size and their own legend settings** (e.g., one subdiagram could show its categories in the legend, whereas another one could hide its own categories).

Subdiagram **ordering** is given by item ordering in the Stacked Diagram page's list. In the future, users could re-arrange subdiagram ordering by dragging and dropping subdiagrams (left as follow-up).

Note there are some shared settings like placement **settings** (e.g., over point), **that the stacked diagram will take from its first subdiagram**. Thus, other subdiagrams won't be able to set or change those specific settings, unless they become the first subdiagram in the stacked diagram. Therefore, when editing a subdiagram that is not located in the first place, we show labels to let users know why they don't see the expected widgets (e.g., placement ones) and where they can find them (i.e., in the first subdiagram).


**GUI Stacked Diagrams**

https://github.com/user-attachments/assets/7b5a55af-45cd-4231-85b9-c6031b783a3f

**GUI Single Diagram**

Before / After ([click](https://downloads.tuxfamily.org/tuxgis/tmp/borrar/single_diagrams_before_after.png) to see it in large size)
![image](https://github.com/user-attachments/assets/d29374f3-9633-4e2c-ba16-d62cd029fea5)

![image](https://github.com/user-attachments/assets/d0573d6b-b1d5-40a7-82b3-c619d1328d46)

**Other samples** 
(Note the corresponding point feature's geometry)

![expected_stackedhistograms](https://github.com/user-attachments/assets/af8539e6-33f0-4cdf-8973-22e1d0bd3511)

![expected_stackedpieshorizontalwithspacing](https://github.com/user-attachments/assets/ece81fd9-f110-45ac-a41b-b7b70b834f81)

![image](https://github.com/user-attachments/assets/92e4d2b3-2916-4942-bdc9-70afb4de59ac)

![image](https://github.com/user-attachments/assets/e7a3f7bd-d7a2-4cfc-b231-1b078f63719f)

Mixed renderer type (pies using single category and histograms using linearly interpolated one):
![expected_stackedpiehistogram](https://github.com/user-attachments/assets/f11c4e10-4a9e-401b-9b58-8848da8132c9)


Nested stacked diagrams (just because we can! Not exposed through GUI, but possible in core and showcased via unit test):
![expected_stackeddiagramsnested](https://github.com/user-attachments/assets/f1eef64d-d337-4693-aa35-a80874348595)

-------------
Includes tests.

Additional fixes:
 + Fix #58501
 + Fix #58782 

----------

Funded by [Landesamt für Vermessung und Geoinformation (LVG) Vorarlberg](https://vorarlberg.at/-/landesamt-fuer-vermessung-und-geoinformation-l-1) in collaboration with the [QGIS user group Switzerland](https://qgis.ch/).

gacarrillor committed:
Introduce QgsStackedDiagram to create compound diagrams like population pyramids. Use Stacked as internal name for stacked diagrams and StackedBar for stacked bar diagrams. Add subDiagram parameter to QgsDiagramRenderer::diagramSize() and QgsDiagramRenderer::diagramSettings() (single and interpolated) to get subdiagrams info.

gacarrillor committed:
[tests] Add first test for stacked diagrams (population pyramid)

gacarrillor committed:
Replace outdated unused in docs; remove outdated unused in code

gacarrillor committed:
Const method; improve new QgsStackedDiagram's docs

gacarrillor committed:
Adjust documentation links

gacarrillor committed:
[tests] Add test for stacked diagrams with spacing

gacarrillor committed:
[tests] Add test for vertical stacked (pie) diagrams with fixed size

gacarrillor committed:
[tests] Fix vertical position for horizontally stacked diagrams and viceversa

gacarrillor committed:
[tests] Subdiagram position in vertically stacked diagrams (iterating subdiagrams backwards)

gacarrillor committed:
[tests] Vertically stacked diagrams with spacing

gacarrillor committed:
[gui] Introduce StackedDiagramProperties and use it as dialog and as vector properties page. For the moment, it works with single diagrams, hiding controls for stacked diagrams

gacarrillor committed:
[gui] Add methods to add and remove subdiagrams in QgsStackedDiagramProperties; remove subdiagram button disabled when subdiagram count is 2 (minimum value to have a stacked diagram); when switching from stacked to single diagram, extra tabs are hidden

gacarrillor committed:
[gui] Implement apply() for QgsStackedDiagramProperties, delegating parts to QgsDiagramProperties for subdiagrams

gacarrillor committed:
[gui] In QgsStackedDiagramProperties, make tabs movable to redefine subdiagrams order

gacarrillor committed:
[fix] Fix issue on histogram bar orientation in a first set up (fix #58501)

gacarrillor committed:
[gui] Move 'No diagram' option from QgsDiagramProperties to QgsStackedDiagramProperties

gacarrillor committed:
[gui] QgsStackedDiagramProperties: Move EngineSettingsButton from QgsDiagramProperties to the side of the main diagram type combobox

gacarrillor committed:
[gui] Implement syncToLayer for QgsStackedDiagramproperties, delegating to subdiagrams' syncToLayer when needed

gacarrillor committed:
[gui] Write/read stacked diagram state to/from XML (QGIS project)

gacarrillor committed:
[core] Introduce QgsStackedDiagramRenderer for rendering stacked diagrams. This allows subdiagrams to define their own sizes and legends; add tests for a stacked diagram composed of mixed renderers (single category and linearly interpolated categories), as well as nested stacked diagrams.

nyalldawson commented:
Nice work @gacarrillor  !

I don't think the use of tabs is appropriate here for configuring sub diagrams. I'd rather see the UI modeled off our existing rule based labeling approach, which IMO is a better fit for this kind of configuration.

nyalldawson commented:
I'd also suggest that the "stacked diagram" option should appear in the topmost combo box alongside the actual diagram types (eg alongside Histogram etc). Then there's no added GUI complexity when using the common single diagram type, and selecting "stacked diagrams" would expose the list of subdiagrams in the panel below that combo.

andreasneumann commented:
> Funded by the [QGIS user group Switzerland](https://qgis.ch/).

It is a co-funding between "Landesvermessungsamt Vorarlberg" (the larger share) and the Swiss QGIS user group (the smaller share). So actually, "Landesvermessungsamt Vorarlberg"  should be listed first.

andreasneumann commented:
Looking forward to use this in the future - a really nice extension to the diagram system!

gacarrillor commented:
> It is a co-funding between "Landesvermessungsamt Vorarlberg" (the larger share) and the Swiss QGIS user group (the smaller share). So actually, "Landesvermessungsamt Vorarlberg" should be listed first.

@andreasneumann, would

Funded by [Landesamt für Vermessung und Geoinformation (LVG)](https://vorarlberg.at/-/landesamt-fuer-vermessung-und-geoinformation-l-1) in collaboration with the [QGIS user group Switzerland](https://qgis.ch/).

work for you or should we stick to 

Funded by [Landesvermessungsamt Vorarlberg](https://vorarlberg.at/-/landesamt-fuer-vermessung-und-geoinformation-l-1) in collaboration with the [QGIS user group Switzerland](https://qgis.ch/).

or another one that you prefer?

gacarrillor commented:
Thank you all for the feedback!
I'm excited about this feature!

andreasneumann commented:
> Funded by [Landesvermessungsamt Vorarlberg](https://vorarlberg.at/-/landesamt-fuer-vermessung-und-geoinformation-l-1) in collaboration with the [QGIS user group Switzerland](https://qgis.ch/).

Officially it is called "Landesamt für Vermessung und Geoinformation (LVG) Vorarlberg". I think Vorarlberg should be in the name. It should also be used in the visual changelog. Thanks a lot!

m-kuhn commented on src/core/qgsdiagramrenderer.cpp:
```diff
@@ -804,6 +837,255 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+QList< QgsLayerTreeModelLegendNode * > QgsStackedDiagramRenderer::legendItems( QgsLayerTreeLayer *nodeLayer ) const
+{
+  QList< QgsLayerTreeModelLegendNode * > nodes;
+  for ( int i = 0; i < rendererCount(); i++ )
```
Any reason not to use a range based for loop here?

`for( QgsDiagramRenderer *renderer : mDiagramRenderers )`?

gacarrillor replied:
Not really @m-kuhn, I could adjust it.

m-kuhn replied:
They are normally preferable

gacarrillor replied:
Addressed in https://github.com/qgis/QGIS/commit/f81df2876c9874f251eb01d7551f443fa37e2df3#diff-c2267f7efae558b8c9bc33e2e751364580eb0b84a7b9b709d3ac815b601bf8eaR977

Is this form also welcome?

`for ( const auto &renderer : std::as_const( mDiagramRenderers ) )`

m-kuhn commented:
@gacarrillor is there anything that keeps this in "draft"? Feature freeze is in two days and this would be a fantastic feature to ship with the next LTR, so I wonder if we have any chance to see this there.

gacarrillor commented:
> @gacarrillor is there anything that keeps this in "draft"? Feature freeze is in two days and this would be a fantastic feature to ship with the next LTR, so I wonder if we have any chance to see this there.

Thanks for pointing this out. I've just removed the draft mark.
I'm currently addressing the GUI/UX suggestions made by @nyalldawson, which are really convenient here.
Looking forward to seeing this landing in QGIS 3.40.

gacarrillor commented:
Thank you @nyalldawson for the exemption.
While addressing your review, we felt we had a great opportunity to finally move the diagram dialog to the Style dock widget, and we're really excited about bringing this to QGIS (we're doing our best). I hope we can push by mid next week.

Preview:
![image](https://github.com/user-attachments/assets/89fd8610-e5e6-41b6-ae27-4aba9090f8f3)

gacarrillor committed:
[gui] Move Diagrams properties to a panel widget, showing tabs instead of a list when docked; harmonize stacked diagram configuration with rule-based labeling; allow QgsDiagramProperties to sync to renderers and diagram layer settings, in addition to syncing to layers; when editing a subdiagram of a stacked diagram, only show widgets for diagram layer settings if the subdiagram is the first one, for the rest, hide those widgets and show a note informing users; make sure stacked diagrams handle enabled and disabled subdiagrams (i.e., don't take into account disabled subdiagrams) and add a test for it; switching from single to stacked diagram: take the single diagram definition as the first stacked diagram; fix #58782 (calling twice the apply method for label rendering)

gacarrillor commented:
Demo:

https://github.com/user-attachments/assets/7b5a55af-45cd-4231-85b9-c6031b783a3f

Known issues to be addressed in upcoming commits:

 + Scale dependent visibility on stacked diagrams.
 + Remove a couple of horizontal scroll bars in dock mode.
 + Replace remaining dialogs in dock widget mode.
 + GUI/UX unexpected behaviors, like default values/default selections.

Most-likely for a follow-up:

 + Drag and drop for diagrams.
 + Copy/paste of diagrams.
 + Passing from stacked to single diagram: take first stacked diagram as the single one.

-------

@nyalldawson, there are still some details to be solved, but we've almost finished addressing your suggestions, thanks!

gacarrillor committed:
[core] Stacked diagrams: make sure scale dependent visibility works on subdiagrams; test included

gacarrillor committed:
[gui] Diagrams: Promote 'increase small diagrams' check box to group box to remove horizontal scroll bar in panel; bring back default values that had been lost in recent changes

3nids commented on src/core/diagram/qgsstackedbardiagram.h:
```diff
@@ -16,7 +16,7 @@
 #define QGSSTACKEDBARDIAGRAM_H
 
-#define DIAGRAM_NAME_STACKED "Stacked"
+#define DIAGRAM_NAME_STACKED_BAR "StackedBar"
```
I am not 100% sure, but we usually prefer static const members over #define
https://stackoverflow.com/questions/1637332/static-const-vs-define
Might be worth switching them?

gacarrillor replied:
Yes, definitely worth it! Thanks for the suggestion.
I'll address this one tomorrow.

gacarrillor replied:
Done in https://github.com/qgis/QGIS/pull/58568/commits/463bf249bb04c91f6cadd5cab834fd3ee836e446

3nids commented on src/core/diagram/qgsstackeddiagram.cpp:
```diff
@@ -0,0 +1,79 @@
+  Q_UNUSED( attributes )
+  Q_UNUSED( c )
+  Q_UNUSED( s )
+  return QSizeF( 0, 0 );
```
Do you know why we have `diagramSize` implemented both in `QgsStackedDiagram` and `QgsStackedDiagramRenderer` ?

gacarrillor replied:
It's the diagram who calculates the diagram size for a specific feature, but depending on the rendering type, the rendering subclass needs to pass additional arguments to the diagram to do that.

3nids commented on src/core/qgsdiagramrenderer.cpp:
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+
+  for ( const auto &stackedRenderer : stackedRenderers )
+  {
+    if ( stackedRenderer->rendererName() == QStringLiteral( "Stacked" ) )
```
stacked diagrams in stacked diagrams
![inception-kukkendare](https://github.com/user-attachments/assets/426a4eb6-2030-4b0a-a1b1-809822a9f9cb)

gacarrillor replied:
When I first thought about it, and gave it a try, that specific `if` clause was the only missing piece.

Note this opens the opportunity to have grids of diagrams instead of _only_ vertical or horizontal stacks.

As mentioned in the PR description, this functionality is not yet offered via GUI, but only demonstrated via unit test.

3nids commented on src/core/qgsdiagramrenderer.cpp:
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+{
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  return QSizeF( 0, 0 );
```
Why is this does needed to go through the children and return the global size?

gacarrillor replied:
`diagramSize` is protected, it'll only be called inside the public method `sizeMapUnits()`, which we do override for Stack Diagrams, delegating to subrenderers and adding up their sizes to get the stacked diagram size.

So, for the Stacked Diagram Renderer, the `diagramSize()` is never called, and that's why we do a dummy implementation returning a default value (it's a pure virtual method, so we needed to override it).

m-kuhn replied:
In this case, would it make sense to make it `private` instead of `protected`?

gacarrillor replied:
Yeap, it's existing API. It's in the abstract superclass, it must be overridden by subclasses.

3nids commented on src/core/qgsdiagramrenderer.cpp:
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+  Q_UNUSED( feature )
+  Q_UNUSED( c )
+  Q_UNUSED( s )
+  return false;
```
This doesn't sound super nice to implement the methods here. But I don't see any approach to avoid this.

gacarrillor replied:
Yeap, those are pure virtual methods, so I was required to override them returning a default value, but they're never actually called for the stacked diagram renderer. That renderer is _special_ in that sense, it delegates several of their methods to sub renderers.

3nids commented on src/core/qgsdiagramrenderer.cpp:
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+    if ( s.scaleBasedVisibility )
+    {
+      double maxScale = s.maximumScale;
+      if ( maxScale > 0 && c.rendererScale() < maxScale )
```
isn't one the boundaries inclusive for scale visibility?

gacarrillor replied:
Good catch! I think you've found a good source of unexpected behaviors. I had copied this block from the current `QgsVectorLayerDiagramProvider::registerDiagram()`.

After some GUI tests with a given scale (say 1:50.000), the scale is sometimes a round number and sometimes a number like `49999.999999998996` or `50000.000000003354` (I could reproduce it when maximizing/restoring the QGIS main window), so those checks were failing from time to time, which could make the diagrams disappear when they're expected to be rendered and vice versa. 

I added checks using qgsDoubleNear() to take non-round values into account.
Applied in https://github.com/qgis/QGIS/pull/58568/commits/b8f976012ab383e29e7156ff75ea3ea836b99e06

3nids commented on src/core/qgsdiagramrenderer.cpp:
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+QList<QgsDiagramSettings> QgsStackedDiagramRenderer::diagramSettings() const
+{
+  QList<QgsDiagramSettings> settingsList;
+  settingsList.push_back( mSettings );
```
why do you need a list of settings and not just the settings?

gacarrillor replied:
That's the way the superclass is implemented (Pure virtual method as well). So far, subclasses stick to have a list of one settings element.

The rationale is probably to have the possibility to somehow extend the renderer to have more available settings.

3nids commented on src/core/qgsdiagramrenderer.cpp:
```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+  {
+    // We draw vertical diagrams backwards, so
+    // we return the subrenderers in reverse order
+    QList< QgsDiagramRenderer * >::const_reverse_iterator iter = mDiagramRenderers.rbegin();
```
you can use a oneliner
`std::reverse(mDiagramRenderers.begin(), mDiagramRenderers.end());`

gacarrillor replied:
Nice tip, thanks! Looks cleaner now.
I've used it in a copy instead, to avoid getting confused on what is the (normal) first diagram after switching modes (e-g-. from vertical to horizontal).

3nids reviewed (commented):
I looked at the core part. I will do the gui tomorrow!
Looks good so far!

gacarrillor committed:
[gui] Diagrams panel: make sure color and effects buttons can open a subpanel

gacarrillor committed:
[core] Stacked diagrams: use one-liner to reverse renderers (std::reverse)

3nids commented on src/gui/vector/qgsdiagramproperties.cpp:
```diff
@@ -1118,3 +1245,71 @@ void QgsDiagramProperties::createAuxiliaryField()
+  {
+    if ( QgsSymbolButton *w = qobject_cast<QgsSymbolButton *>( widget ) )
+    {
+      connect( w, SIGNAL( changed() ), this, signal );
```
Can you switch theses to functor based connections?

gacarrillor replied:
Yes, should I also adjust https://github.com/qgis/QGIS/blob/da1f55b7d1ca1026535ebadb39303df4447cc729/src/gui/symbology/qgsrendererpropertiesdialog.cpp#L140 and https://github.com/qgis/QGIS/blob/da1f55b7d1ca1026535ebadb39303df4447cc729/src/gui/qgstextformatwidget.cpp#L662 for consistency?

gacarrillor replied:
Done in https://github.com/qgis/QGIS/pull/58568/commits/01055d7a2d48f999ea193410fec7e5764d5135cc and https://github.com/qgis/QGIS/pull/58568/commits/eb201c56e61919844b21659f9c3493d3a6441332

3nids reviewed (approved):
Not much more comments regarding the GUI, good job!

gacarrillor commented:
@DelazJ, @nirvn, I'd love your feedback on the GUI/UX part if you can find some time to check it.

gacarrillor committed:
[core] Address review: switch from define to static const

gacarrillor committed:
[gui] Diagram panel: connect effects and font widgets to live updates; switch from string-based connections to functor-based ones

gacarrillor committed:
[gui] Switch from string-based connections to functor-based ones

3nids commented on src/core/diagram/qgspiediagram.h:
```diff
@@ -38,6 +36,8 @@ class QgsRenderContext;
 class CORE_EXPORT QgsPieDiagram: public QgsDiagram SIP_NODEFAULTCTORS
 {
   public:
+    static const QString DIAGRAM_NAME_PIE;
```
```suggestion
    static const QString DIAGRAM_NAME_PIE SIP_SKIP;
```
can you do the same for all the others?

gacarrillor committed:
[core] Addres review: skip SIP for some static const diagram names

gacarrillor committed:
[core] Address review, stacked diagrams: include maxScale for rendering with scale dependent visibility

gacarrillor commented:
Thanks a lot @3nids for your insightful review!

nyalldawson commented on src/core/diagram/qgsstackedbardiagram.cpp:
```diff
@@ -117,7 +119,7 @@ double QgsStackedBarDiagram::legendSize( double value, const QgsDiagramSettings
 QString QgsStackedBarDiagram::diagramName() const
 {
-  return DIAGRAM_NAME_STACKED;
+  return QgsStackedBarDiagram::DIAGRAM_NAME_STACKED_BAR;
```
@gacarrillor this has broken restoring old projects using stacked bar diagrams -- you'll need to revert this change, and then introduce a new unique, different string for the stacked diagram type.

gacarrillor replied:
Thanks for the report and sorry.
Solved in #58915

nyalldawson commented on src/core/qgsdiagramrenderer.cpp:
```diff
@@ -804,6 +857,273 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+{
+  if ( renderer )
+  {
+    mDiagramRenderers.append( renderer );
```
@gacarrillor What is supposed to happen with ownership here? Right now mDiagramRenderers is never cleaned up, so these are all leaked. There's also no SIP_TRANSFER annotation on the function declaration, so you'll get a crash if you call this method from python.

There's also a (very bad) ownership issue with clone() here, because that calls a default copy constructor which just directly copies mDiagramRenderers to the clone, even though all the pointers in that container belong to the original instance. 

Can you please address this ASAP? We are closing in on release and this API is broken in its current state...

gacarrillor replied:
I'll have a look at it, thanks a lot for pointing this out!

gacarrillor replied:
Addressed in https://github.com/qgis/QGIS/pull/59123
